### DISCUSSION TITLE
[@types/cookie-parser]- cookies of type 'any': unsafe-member-access / unsafe-assignment

### DISCUSSION THREAD

**Discussion by adamwozhere** [DefinitelyTyped/DefinitelyTyped#68587] [Issues with a @types package] [answered]
In my Express project I'm using cookie-parser to use a jwt refresh token in a cookie. In my RequestHandler I am attempting to desctucture cookies `const { cookies } = req;` but @typescript/eslint errors `Unsafe array destructuring of tuple element with an "any" value.` this also stops me from accessing `req.cookies?.jwt` as cookies is of type `any`.

Apologies if I am misunderstanding: I am currently extending the Request interface to get around this problem, but shouldn't cookies be a typed property in `@types/cookie-parser` ? As:

```
interface Request {
   cookies?: { [key: string]: string };
}
```

[a comment by an automated account is not shown]

**Comment by MysteryBlokHed** [comment]
Could you provide a larger code snippet that shows this problem?

**Reply by adamwozhere** [reply]
Sure, here is my request handler:
```
import { Request, Response } from 'express';

export const logoutHandler = (req: Request, res: Response) => {
  const { cookies } = req;

  if (!cookies?.jwt) {
    return res.sendStatus(204);
  }

  // TODO: remove jwt refresh token from DB

  res.clearCookie('jwt', { httpOnly: true, sameSite: 'none', secure: true });
  return res.sendStatus(204);
};
```
And the errors showing in VScode:
![request_error-1](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/96954293/625c3991-f8cc-4a72-a7af-26b0ecbbf632)
![request_error-2](https://github.com/DefinitelyTyped/DefinitelyTyped/assets/96954293/7e70072d-f5f4-428a-977d-8ee058404d5f)

**Reply by MysteryBlokHed** [reply]
I just opened #68591 to fix this

**Reply by MysteryBlokHed** [reply]
It looks like it might actually be problematic to try and type `cookies` and `signedCookies`. Since the Express `Request` type defines both of these properties as `any`, I don't know whether there's actually a way to override them with a narrower type.

The best solution might just be to use a type assertion (i.e. `const cookies = req.cookies as Record<string, string>`), which TypeScript should be fine with since the original type is `any`.

**Reply by adamwozhere** [reply]
Thanks for this - I'm generally new to GitHub / opensource and Typescript. I had tried adding cookies to the index.d.ts file under the `secret?: string;` but it didn't work: I can see in your issue you have added the `req.cookies = Object.create(null)` so I am guessing that the Express properties (`cookies: any`) can't be overridden but the `req.cookies` in cookie-parser can (if it's added to the library) ?

I wonder if it's possible to add a generic type to the Express Request in [@types/express-serve-static-core] for the cookies so they can be typed as you would with the body?
``` 
export interface Request<
    P = ParamsDictionary,
    ResBody = any,
    ReqBody = any,
    ReqQuery = ParsedQs,
    LocalsObj extends Record<string, any> = Record<string, any>,
    Cookies: extends Record<string, any> = Record<string, any>,
> extends http.IncomingMessage, Express.Request {
```
e.g.
```
const handler = (req: Request<object, object, { username: string, password: string}, object, object, { jwt: string }>, res: Response) => {
    const { jwt } = req.cookies;
    //...
}
```

**Reply by MysteryBlokHed** [reply]
The `Request` interface defined in the cookie-parser package is an example of [module augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation); it's technically the *exact same* interface as the Express one. Because of that, I don't think it's possible to give it a narrower type.

It's also not possible to add a generic to it afaik, since that would also conflict with the original type. It would have to be added to the original library defining the `Request` type, which might cause compatibility problems and likely wouldn't be merged.

It's not very elegant, but imo the best solution is just to add type assertions in your own code.

**Reply by adamwozhere** [reply]
Brilliant thanks, I think I understand it now. I decided to try the module augmentation approach instead of the `as Record<string, string>` - I've added a declaration file in my project root and everything works as expected:

```
/* eslint-disable-next-line @typescript-eslint/no-unused-vars */
import * as express from 'express';

declare module 'express' {
  interface Request {
    cookies?: { [key: string]: string };
  }
}
```

I'm using the cookies in a few different places, and as I know it's not going to cause any side-effects in my code, I thought it would be a good solution.

**Reply by MysteryBlokHed** [reply, marked as answer]
Actually, I made a mistake: it looks like there was just an issue with how the cookie-parser types were doing their augmentation. #68591 is now properly setting the types to `Record<string, any>` for `cookies` and `signedCookies`.

If you prefer to use the narrower type (`string` instead of `any`) for your own use case, then you can also keep your own augmentation in your code.

**Reply by adamwozhere** [reply]
I originally thought I had made a mistake with my own project module settings as I tried to declare namespace as in cookie-parser but it didn't work - so it's interesting to see that it is also now declare module! Thank you for sorting this!
